### ISSUE TITLE
Canal - Network Policy only works when both application are on same node

### ISSUE THREAD

**Issue by msschl** [projectcalico/calico#2971]
I am using a Kubernetes NetworkPolicy to allow my public nginx-ingress-controller to egress to all pods in all namespaces that have the labels ingress="true" and ingress-controller="public". But strangely it only works when the applications are on the same node as the nginx-ingress-controller

This issue seems related to #2898 or #2169

```yaml
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: allow-all-traffic-to-public-web-apps
  namespace: public
spec:
  podSelector:
    matchLabels:
      app: "nginx-ingress"
      component: "controller"
      release: "public-nginx-ingress"
  policyTypes:
  - Egress
  egress:
  - to:
    - namespaceSelector: {}
      podSelector:
        matchLabels:
          ingress-controller: "public"
          allow-ingress: "true"
```

## Expected Behavior
When I apply the network policy, regardless of the Kubernetes worker node, the policy should work.

## Current Behavior
When I apply the network policy, it only works if the application that that the nginx-ingress-controller is connecting to is on same Kubernetes worker node.

## Your Environment
```sh
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.7", GitCommit:"8fca2ec50a6133511b771a11559e24191b1aa2b4", GitTreeState:"clean", BuildDate:"2019-09-18T14:47:22Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:09:08Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```
canal:
```yaml
kubectl describe ds canal -n kube-system
Name:           canal
Selector:       k8s-app=canal
Node-Selector:  beta.kubernetes.io/os=linux
Labels:         k8s-app=canal
Annotations:    deprecated.daemonset.template.generation: 1
                kubectl.kubernetes.io/last-applied-configuration:
                  {"apiVersion":"apps/v1","kind":"DaemonSet","metadata":{"annotations":{},"labels":{"k8s-app":"canal"},"name":"canal","namespace":"kube-syst...
Desired Number of Nodes Scheduled: 3
Current Number of Nodes Scheduled: 3
Number of Nodes Scheduled with Up-to-date Pods: 3
Number of Nodes Scheduled with Available Pods: 3
Number of Nodes Misscheduled: 0
Pods Status:  3 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:           k8s-app=canal
  Annotations:      scheduler.alpha.kubernetes.io/critical-pod:
  Service Account:  canal
  Init Containers:
   install-cni:
    Image:      calico/cni:v3.10.0
    Port:       <none>
    Host Port:  <none>
    Command:
      /install-cni.sh
    Environment:
      CNI_CONF_NAME:         10-canal.conflist
      CNI_NETWORK_CONFIG:    <set to the key 'cni_network_config' of config map 'canal-config'>  Optional: false
      KUBERNETES_NODE_NAME:   (v1:spec.nodeName)
      SLEEP:                 false
    Mounts:
      /host/etc/cni/net.d from cni-net-dir (rw)
      /host/opt/cni/bin from cni-bin-dir (rw)
   flexvol-driver:
    Image:        calico/pod2daemon-flexvol:v3.10.0
    Port:         <none>
    Host Port:    <none>
    Environment:  <none>
    Mounts:
      /host/driver from flexvol-driver-host (rw)
  Containers:
   calico-node:
    Image:      calico/node:v3.10.0
    Port:       <none>
    Host Port:  <none>
    Requests:
      cpu:      250m
    Liveness:   exec [/bin/calico-node -felix-live] delay=10s timeout=1s period=10s #success=1 #failure=6
    Readiness:  http-get http://localhost:9099/readiness delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      DATASTORE_TYPE:                     kubernetes
      USE_POD_CIDR:                       true
      WAIT_FOR_DATASTORE:                 true
      NODENAME:                            (v1:spec.nodeName)
      CALICO_NETWORKING_BACKEND:          none
      CLUSTER_TYPE:                       k8s,canal
      FELIX_IPTABLESREFRESHINTERVAL:      60
      IP:
      CALICO_IPV4POOL_CIDR:               192.168.0.0/16
      CALICO_DISABLE_FILE_LOGGING:        true
      FELIX_DEFAULTENDPOINTTOHOSTACTION:  ACCEPT
      FELIX_IPV6SUPPORT:                  false
      FELIX_LOGSEVERITYSCREEN:            info
      FELIX_HEALTHENABLED:                true
    Mounts:
      /lib/modules from lib-modules (ro)
      /run/xtables.lock from xtables-lock (rw)
      /var/lib/calico from var-lib-calico (rw)
      /var/run/calico from var-run-calico (rw)
      /var/run/nodeagent from policysync (rw)
   kube-flannel:
    Image:      quay.io/coreos/flannel:v0.11.0
    Port:       <none>
    Host Port:  <none>
    Command:
      /opt/bin/flanneld
      --ip-masq
      --kube-subnet-mgr
    Environment:
      POD_NAME:           (v1:metadata.name)
      POD_NAMESPACE:      (v1:metadata.namespace)
      FLANNELD_IFACE:    <set to the key 'canal_iface' of config map 'canal-config'>  Optional: false
      FLANNELD_IP_MASQ:  <set to the key 'masquerade' of config map 'canal-config'>   Optional: false
    Mounts:
      /etc/kube-flannel/ from flannel-cfg (rw)
      /run/xtables.lock from xtables-lock (rw)
  Volumes:
   lib-modules:
    Type:          HostPath (bare host directory volume)
    Path:          /lib/modules
    HostPathType:
   var-run-calico:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/calico
    HostPathType:
   var-lib-calico:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/calico
    HostPathType:
   xtables-lock:
    Type:          HostPath (bare host directory volume)
    Path:          /run/xtables.lock
    HostPathType:  FileOrCreate
   flannel-cfg:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      canal-config
    Optional:  false
   cni-bin-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /opt/cni/bin
    HostPathType:
   cni-net-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/cni/net.d
    HostPathType:
   policysync:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/nodeagent
    HostPathType:  DirectoryOrCreate
   flexvol-driver-host:
    Type:          HostPath (bare host directory volume)
    Path:          /usr/libexec/kubernetes/kubelet-plugins/volume/exec/nodeagent~uds
    HostPathType:  DirectoryOrCreate
Events:
  Type    Reason            Age   From                  Message
  ----    ------            ----  ----                  -------
  Normal  SuccessfulCreate  14m   daemonset-controller  Created pod: canal-vrkbv
```

canal is working correctly. no crash or restarts
```sh
$ kubectl get pods -n kube-system -o wide
NAME                                     READY   STATUS    RESTARTS   AGE     IP            NODE             NOMINATED NODE   READINESS GATES
canal-9d5dl                              2/2     Running   0          2d19h   172.18.2.91   lxkubemaster01   <none>           <none>
canal-pxgnw                              2/2     Running   0          21h     172.18.2.92   lxkubenode01     <none>           <none>
canal-vrkbv                              2/2     Running   0          16m     172.18.2.93   lxkubenode02     <none>           <none>
coredns-6c5cb5bb5f-7mndq                 1/1     Running   0          2d19h   10.244.1.16   lxkubenode01     <none>           <none>
coredns-6c5cb5bb5f-fz49s                 1/1     Running   0          2d19h   10.244.2.27   lxkubenode02     <none>           <none>
etcd-lxkubemaster01                      1/1     Running   0          2d19h   172.18.2.91   lxkubemaster01   <none>           <none>
kube-apiserver-lxkubemaster01            1/1     Running   0          2d19h   172.18.2.91   lxkubemaster01   <none>           <none>
kube-controller-manager-lxkubemaster01   1/1     Running   2          2d19h   172.18.2.91   lxkubemaster01   <none>           <none>
kube-proxy-42mqr                         1/1     Running   0          2d19h   172.18.2.92   lxkubenode01     <none>           <none>
kube-proxy-7fr5z                         1/1     Running   0          2d19h   172.18.2.93   lxkubenode02     <none>           <none>
kube-proxy-grxwp                         1/1     Running   0          2d19h   172.18.2.91   lxkubemaster01   <none>           <none>
kube-scheduler-lxkubemaster01            1/1     Running   2          2d19h   172.18.2.91   lxkubemaster01   <none>           <none>
```

I am noticing a small difference in the iptabels from node 01 to node 02
lxkubenode02 is missing entry `CNI portfwd requiring masquerade`
```sh
lxkubenode01:~$ iptables --list POSTROUTING -n -t nat -v
Chain POSTROUTING (policy ACCEPT 67 packets, 5601 bytes)
 pkts bytes target     prot opt in     out     source               destination
 761K   73M cali-POSTROUTING  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* cali:O3lYWMrLQYEMJtB5 */
 761K   73M CNI-HOSTPORT-MASQ  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* CNI portfwd requiring masquerade */
3175K  289M KUBE-POSTROUTING  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes postrouting rules */
 7099  939K RETURN     all  --  *      *       10.244.0.0/16        10.244.0.0/16
 103K 6182K MASQUERADE  all  --  *      *       10.244.0.0/16       !224.0.0.0/4
 148K 8857K RETURN     all  --  *      *      !10.244.0.0/16        10.244.1.0/24
    0     0 MASQUERADE  all  --  *      *      !10.244.0.0/16        10.244.0.0/16

lxkubenode02:~$ iptables --list POSTROUTING -n -t nat -v
Chain POSTROUTING (policy ACCEPT 90 packets, 8485 bytes)
 pkts bytes target     prot opt in     out     source               destination
3111K  285M cali-POSTROUTING  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* cali:O3lYWMrLQYEMJtB5 */
3121K  286M KUBE-POSTROUTING  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes postrouting rules */
 6787  920K RETURN     all  --  *      *       10.244.0.0/16        10.244.0.0/16
 103K 6182K MASQUERADE  all  --  *      *       10.244.0.0/16       !224.0.0.0/4
 190K   11M RETURN     all  --  *      *      !10.244.0.0/16        10.244.2.0/24
    0     0 MASQUERADE  all  --  *      *      !10.244.0.0/16        10.244.0.0/16
```

I am using a global-deny-all rule:
```yaml
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
metadata:
  name: global-deny-all
spec:
  order: 2000
  types:
  - Ingress
  - Egress

  # Egress network rules
  egress:
  # Allow all egress traffic from kube-system to cluster internals
  - action: Allow
    source:
      namespaceSelector: name == 'kube-system'
    destination: {}

  # Allow egress DNS traffic from coredns to 172.18.2.21 (company internal dns server)
  - action: Allow
    protocol: UDP
    source:
      selector: k8s-app == 'kube-dns'
      namespaceSelector: name == 'kube-system'
    destination:
      nets:
        - 172.18.2.21/32
      ports:
        - 53

  # Ingress network rules
  ingress:
  # Allow all ingress traffic from outside to the kube-system namespace.
  - action: Allow
    source: {}
    destination:
      namespaceSelector: name == 'kube-system'
```

**Comment by tmjd** [comment]
I do not see any policy that would allow ingress to the pods that are suppose to be accepting traffic from the ingress controller. I'm surprised the traffic from the ingress controller is working locally. Is the ingress controller running as hostNetworked? If so then policy doesn't actually apply to it unless you've defined Host Endpoints for your nodes, and could explain why they are able to reach local pods.

**Comment by msschl** [comment]
The ingress controller is bound to the hostNetwork.

**Comment by msschl** [comment]
As I understand it, I have an egress policy which allows the ingress controller to connect to any pod in any namespace with the podLabels `ingress-controller: "public" ` and `allow-ingress: "true"`. Therefor the ingress controller should be able to connect to the pods.

**Comment by tmjd** [comment]
Like I mentioned, policy does not apply to hostNetworked pods so any policy written to select the ingress pod will not work.

Even with that said because you have a deny all policy that applies for ingress traffic, so you will need policy that applies to your destination pods that allows ingress to them.
Lets assume that your current (source) pod trying access your application is pod networked (not hostNetworked), even if you applied egress policy to your source pod, if your application pod has policy that denies ingress traffic (which your default deny does) then there must be explicit policy that allows ingress.

In other words, just because you've allowed pod A to send traffic to pod B, if pod B is not allowed to receive traffic from pod A then the communication is blocked.

**Comment by mohideen** [comment]
I had a similar issue on RHEL 8 which was resolved when I rebootrapped the cluster with the latest canal.yaml from calico. Also, I added the `FELIX_IPTABLESBACKEND` to the canal daemonset container spec env list.

```
           # Set IPTABLE backend to auto
            - name: FELIX_IPTABLESBACKEND
              value: "auto"
```
But, I'm of the impression that `auto` is the default option for that configuration now.

**Comment by mohideen** [comment]
Actually, it only worked temporarily and it stopped working now. I see the same network timeout issue when pods across nodes try to talk to each other. I think it stopped working after we restarted our nodes (VMs) for weekly maintenance.

I do suspect that the source IP is getting changed from the original pod's IP to a different IP when it reached the target node. I can get the pods to talk across nodes if I change the policy to use ipBlock of the podSubnet CIDR.

Desired Policy (not working)
```
# Allow all traffic between app-stack pods
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: app-stack-allow-all-interpod
  namespace: default
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/instance: app-stack
  policyTypes:
  - Ingress
  - Egress
  egress:
  - to:
    - podSelector:
        matchLabels:
          app.kubernetes.io/instance: app-stack
  ingress:
  - from:
    - podSelector:
        matchLabels:
          app.kubernetes.io/instance: app-stack
```

Working policy (using ipBlock)
```
# Allow all traffic between app-stack pods
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: app-stack-allow-all-interpod
  namespace: default
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/instance: app-stack
  policyTypes:
  - Ingress
  - Egress
  egress:
  - to:
    - podSelector:
        matchLabels:
          app.kubernetes.io/instance: app-stack
  ingress:
  - from:
    - ipBlock:
        cidr: 10.244.0.0/16
```

And, I did confirm that both the pods have the label `app.kubernetes.io/instance: app-stack`

```
$ kubectl get pods -l app.kubernetes.io/instance=app-stack -owide 
NAME                   READY   STATUS    RESTARTS   AGE    IP            NODE                   NOMINATED NODE   READINESS GATES
app-stack-db-sset-0   1/1     Running   0          154m   10.244.1.33   worker2.mydomain       <none>           <none>
app-stack-sset-0      0/1     Running   0          154m   10.244.2.24   worker3.mydomain       <none>           <none>
```

Can anyone help me troubleshoot this networking issue? Thanks

**Comment by mohideen** [comment]
@tmjd I tried looking into the iptables packet as you suggested in issue #1923. When I diff`ed the packet count before and after, I got the below output:
```
< [0:0] -A cali-fw-calib2b14591c8c -m comment --comment "cali:SAb1o07tWVVNkA8R" -m conntrack --ctstate INVALID -j DROP
---
> [2:80] -A cali-fw-calib2b14591c8c -m comment --comment "cali:SAb1o07tWVVNkA8R" -m conntrack --ctstate INVALID -j DROP
48c48
< [2:80] -A cali-fw-cali2f5211f64be -m comment --comment "cali:zxne0U1qIO3lgFtH" -m conntrack --ctstate INVALID -j DROP
---
> [4:160] -A cali-fw-cali2f5211f64be -m comment --comment "cali:zxne0U1qIO3lgFtH" -m conntrack --ctstate INVALID -j DROP
53c53
< [0:0] -A cali-tw-cali467dcb24199 -m comment --comment "cali:xP8YZ83mJf8zUVRp" -m comment --comment "Drop if no policies passed packet" -m mark --mark 0x0/0x20000 -j DROP
---
> [3:180] -A cali-tw-cali467dcb24199 -m comment --comment "cali:xP8YZ83mJf8zUVRp" -m comment --comment "Drop if no policies passed packet" -m mark --mark 0x0/0x20000 -j DROP
```
I'm not sure how to decode this. Do you have guidance?

**Comment by caseydavenport** [comment]
Reading through this, I think @tmjd's analysis is correct. Pods running with `hostNetwork: true` are not valid targets for network policy label selectors. You'll need to use host endpoints in order to apply to host-networked things. 

My strong suspicion is that your egress policy is blocking the traffic because the host IP is not included in the egress policy selector. This is, unfortunately, working as intended. You'll need to use an exact IP / subnet in your selector, or use host endpoints to select nodes.
